README
This is the branch where the documentation is stored.
I keep this branch in a subdirectory of my master branch.



CHANGELOG
Hand written changelog of important changes.
Use the github page to view the automated git generated brother.

2016/11/12
----------

+ Changed License to LGPLv3 (from GPLv3)

2011/03/07
----------
+ Fixed lyrix.at parser (ctr was incremented always, not only on item found)
+ Speed up of cover and lyrics plugins (Up to 200%)
+ Fixed lyrics:magistrix plugin
+ Fixed lyrics:lyricswiki plugin
+ Fix of DynHelp
+ Fixed overall memory leaks
+ added finalize() call to cleanup plugins

2011/03/09
----------
+ Help text reacts now on "I DONT WANNA COLOR!"
+ added ability to stop searchengine by returning GLYRE_STOP_BY_CB in callback
+ Fixed a bug when using the google plugin and -D
+ Fixed google (and other) returning 404 Pages
+ Added 'tracklist' getter12
+ Fixed all warnings occuring with highger warn level (Default now)

2011/03/13
----------
+ Added examplce.c
+ Removed alldcovers plugin (they have good image, but don't allow API usage for library and desktop applications it seems :-( )
+ Added format check for image plugins. Almost 100% success (== valid images) with this check now
+ Added option (C API) to configure levenshtein fuzzyness
+ went to bed :-)

2011/03/14
----------
+ Fixed a bug that made the itemcounter to be negative. ("Got -21 images!" he happily replied)

2011/03/17
----------
+ Fixed very ugly stringop.c - Many functions should be faster and smoother now.
+ Fixed -w stdout and tracklist. (fwrite didn't recognize \0, as ->size was the duration here)
+ Fixed an ugly bug in photos.c that crashed libglyr if ignoring a URL
+ added provider photos:google
+ reformatted due to trying anjuta as IDE (instead of pure gVim)

2011/03/18
----------
+ Replaced ugly remove_html_tags with a more generic version that works completely in-place
+ Removed a few redundant strlen()s
+ added (non-working) albumlist getter which gets all official albums from a certain artist

2011/03/19
----------
+ made albumlist getter working. Wohoo.
+ Fixed dumb shortcut bug in glyrc
+ Made finalize() of ainfo,similiar,review,tracklist,albumlist generic, i.e. only call the same function

2011/03/24
----------
+ Fixed crash when libcurl put an error to stdout (ouch..)
+ Fixed google names when having spaces (ouch x 2)

2011/03/28
----------
+ Did a lot of fixing regarding the API
+ Fixed "broken" review getter. Still slow; needs more sources.
+ Extended src/example.c to use more functions.
+ Tested API (It works, biaatch!)

2011/03/29
----------
+ Made 'return GLYRE_STOP_BY_CB;' work.
+ removed Gly_infoat, it was only for devs..

2011/05/15
----------
Found some time to work a bit here :-)
+ Added support for google translator (new 'gtrans' getter), lyrics/text can bew translated directly now.
  It is a bit restricted atm. as only 100.000 chars are allowed per day per IP address. But I already 
  requested more. (20x times as much), hopefully it will get through. Also language detecting works.
+ Glyr does now checksumming. This is useful for developers which want to make sure the data is unique.
  For normal users: Well, the duplicate check is now as fast as rmlint :-P
+ getopt_long is now getopt_long_only (no --arguments, just -argument or -a)
+ removed call_direct. It's of no use. Not even for devs.

2011/05/16
----------
+ removed wikipedia.c for ainfo - never worked. Better provide a link and (relation getter) and let the user render the site (too many information would get lost if glyr would need to parse it)

2011/06/14
----------
+ added support for proxy. (only basic libcurl support, should be suffiecient for 99% of all cases)
+ made API *much* slicker & nicer

2011/06/25
----------
- CHANGELOG discontinued, use the git commit history instead


lib/md5.c
/*
**********************************************************************
** md5.c                                                            **
** RSA Data Security, Inc. MD5 Message Digest Algorithm             **
** Created: 2/17/90 RLR                                             **
** Revised: 1/91 SRD,AJ,BSK,JT Reference C Version                  **
**********************************************************************
*/

/*
 **********************************************************************
 ** Copyright (C) 1990, RSA Data Security, Inc. All rights reserved. **
 **                                                                  **
 ** License to copy and use this software is granted provided that   **
 ** it is identified as the "RSA Data Security, Inc. MD5 Message     **
 ** Digest Algorithm" in all material mentioning or referencing this **
 ** software or this function.                                       **
 **                                                                  **
 ** License is also granted to make and use derivative works         **
 ** provided that such works are identified as "derived from the RSA **
 ** Data Security, Inc. MD5 Message Digest Algorithm" in all         **
 ** material mentioning or referencing the derived work.             **
 **                                                                  **
 ** RSA Data Security, Inc. makes no representations concerning      **
 ** either the merchantability of this software or the suitability   **
 ** of this software for any particular purpose.  It is provided "as **
 ** is" without express or implied warranty of any kind.             **
 **                                                                  **
 ** These notices must be retained in any copies of any part of this **
 ** documentation and/or software.                                   **
 **********************************************************************
 */

#include <stdio.h>
#include <string.h>

#include "md5.h"

/* prot */
static void Transform(uint64_t *buf, uint64_t *in);

static unsigned char PADDING[64] =
{
    0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
};

/* F, G and H are basic MD5 functions: selection, majority, parity */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4 */
/* Rotation is separate from addition to prevent recomputation */
#define FF(a, b, c, d, x, s, ac) \
  {(a) += F ((b), (c), (d)) + (x) + (uint64_t)(ac); \
   (a) = ROTATE_LEFT ((a), (s)); \
   (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) \
  {(a) += G ((b), (c), (d)) + (x) + (uint64_t)(ac); \
   (a) = ROTATE_LEFT ((a), (s)); \
   (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) \
  {(a) += H ((b), (c), (d)) + (x) + (uint64_t)(ac); \
   (a) = ROTATE_LEFT ((a), (s)); \
   (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) \
  {(a) += I ((b), (c), (d)) + (x) + (uint64_t)(ac); \
   (a) = ROTATE_LEFT ((a), (s)); \
   (a) += (b); \
  }

void MD5Init (MD5_CTX *mdContext)
{
    mdContext->i[0] = mdContext->i[1] = (uint64_t)0;

    /* Load magic initialization constants.
     */
    mdContext->buf[0] = (uint64_t)0x67452301;
    mdContext->buf[1] = (uint64_t)0xefcdab89;
    mdContext->buf[2] = (uint64_t)0x98badcfe;
    mdContext->buf[3] = (uint64_t)0x10325476;
}

void MD5Update (MD5_CTX *mdContext, unsigned char *inBuf, unsigned int inLen)
{
    uint64_t in[16];
    int mdi;
    unsigned int i, ii;

    /* compute number of bytes mod 64 */
    mdi = (int)((mdContext->i[0] >> 3) & 0x3F);

    /* update number of bits */
    if ((mdContext->i[0] + ((uint64_t)inLen << 3)) < mdContext->i[0])
    {
        mdContext->i[1]++;
    }
    mdContext->i[0] += ((uint64_t)inLen << 3);
    mdContext->i[1] += ((uint64_t)inLen >> 29);

    while (inLen--)
    {
        /* add new character to buffer, increment mdi */
        mdContext->in[mdi++] = *inBuf++;

        /* transform if necessary */
        if (mdi == 0x40)
        {
            for (i = 0, ii = 0; i < 16; i++, ii += 4)
                in[i] = (((uint64_t)mdContext->in[ii+3]) << 24) |
                        (((uint64_t)mdContext->in[ii+2]) << 16) |
                        (((uint64_t)mdContext->in[ii+1]) << 8) |
                        ((uint64_t)mdContext->in[ii]);
            Transform (mdContext->buf, in);
            mdi = 0;
        }
    }
}

void MD5Final(MD5_CTX *mdContext)
{
    uint64_t in[16];
    int mdi;
    unsigned int i, ii;
    unsigned int padLen;

    /* save number of bits */
    in[14] = mdContext->i[0];
    in[15] = mdContext->i[1];

    /* compute number of bytes mod 64 */
    mdi = (int)((mdContext->i[0] >> 3) & 0x3F);

    /* pad out to 56 mod 64 */
    padLen = (mdi < 56) ? (56 - mdi) : (120 - mdi);
    MD5Update (mdContext, PADDING, padLen);

    /* append length in bits and transform */
    for (i = 0, ii = 0; i < 14; i++, ii += 4)
        in[i] = (((uint64_t)mdContext->in[ii+3]) << 24) |
                (((uint64_t)mdContext->in[ii+2]) << 16) |
                (((uint64_t)mdContext->in[ii+1]) << 8) |
                ((uint64_t)mdContext->in[ii]);
    Transform (mdContext->buf, in);

    /* store buffer in digest */
    for (i = 0, ii = 0; i < 4; i++, ii += 4)
    {
        mdContext->digest[ii] = (unsigned char)(mdContext->buf[i] & 0xFF);
        mdContext->digest[ii+1] =
            (unsigned char)((mdContext->buf[i] >> 8) & 0xFF);
        mdContext->digest[ii+2] =
            (unsigned char)((mdContext->buf[i] >> 16) & 0xFF);
        mdContext->digest[ii+3] =
            (unsigned char)((mdContext->buf[i] >> 24) & 0xFF);
    }
}

/* Basic MD5 step. Transform buf based on in.
 */
static void Transform(uint64_t *buf, uint64_t *in)
{
    uint64_t a = buf[0], b = buf[1], c = buf[2], d = buf[3];

    /* Round 1 */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
    FF ( a, b, c, d, in[ 0], S11, 3614090360UL); /* 1 */
    FF ( d, a, b, c, in[ 1], S12, 3905402710UL); /* 2 */
    FF ( c, d, a, b, in[ 2], S13,  606105819UL); /* 3 */
    FF ( b, c, d, a, in[ 3], S14, 3250441966UL); /* 4 */
    FF ( a, b, c, d, in[ 4], S11, 4118548399UL); /* 5 */
    FF ( d, a, b, c, in[ 5], S12, 1200080426UL); /* 6 */
    FF ( c, d, a, b, in[ 6], S13, 2821735955UL); /* 7 */
    FF ( b, c, d, a, in[ 7], S14, 4249261313UL); /* 8 */
    FF ( a, b, c, d, in[ 8], S11, 1770035416UL); /* 9 */
    FF ( d, a, b, c, in[ 9], S12, 2336552879UL); /* 10 */
    FF ( c, d, a, b, in[10], S13, 4294925233UL); /* 11 */
    FF ( b, c, d, a, in[11], S14, 2304563134UL); /* 12 */
    FF ( a, b, c, d, in[12], S11, 1804603682UL); /* 13 */
    FF ( d, a, b, c, in[13], S12, 4254626195UL); /* 14 */
    FF ( c, d, a, b, in[14], S13, 2792965006UL); /* 15 */
    FF ( b, c, d, a, in[15], S14, 1236535329UL); /* 16 */

    /* Round 2 */
#define S21 5
#define S22 9
#define S23 14
#define S24 20
    GG ( a, b, c, d, in[ 1], S21, 4129170786UL); /* 17 */
    GG ( d, a, b, c, in[ 6], S22, 3225465664UL); /* 18 */
    GG ( c, d, a, b, in[11], S23,  643717713UL); /* 19 */
    GG ( b, c, d, a, in[ 0], S24, 3921069994UL); /* 20 */
    GG ( a, b, c, d, in[ 5], S21, 3593408605UL); /* 21 */
    GG ( d, a, b, c, in[10], S22,   38016083UL); /* 22 */
    GG ( c, d, a, b, in[15], S23, 3634488961UL); /* 23 */
    GG ( b, c, d, a, in[ 4], S24, 3889429448UL); /* 24 */
    GG ( a, b, c, d, in[ 9], S21,  568446438UL); /* 25 */
    GG ( d, a, b, c, in[14], S22, 3275163606UL); /* 26 */
    GG ( c, d, a, b, in[ 3], S23, 4107603335UL); /* 27 */
    GG ( b, c, d, a, in[ 8], S24, 1163531501UL); /* 28 */
    GG ( a, b, c, d, in[13], S21, 2850285829UL); /* 29 */
    GG ( d, a, b, c, in[ 2], S22, 4243563512UL); /* 30 */
    GG ( c, d, a, b, in[ 7], S23, 1735328473UL); /* 31 */
    GG ( b, c, d, a, in[12], S24, 2368359562UL); /* 32 */

    /* Round 3 */
#define S31 4
#define S32 11
#define S33 16
#define S34 23
    HH ( a, b, c, d, in[ 5], S31, 4294588738UL); /* 33 */
    HH ( d, a, b, c, in[ 8], S32, 2272392833UL); /* 34 */
    HH ( c, d, a, b, in[11], S33, 1839030562UL); /* 35 */
    HH ( b, c, d, a, in[14], S34, 4259657740UL); /* 36 */
    HH ( a, b, c, d, in[ 1], S31, 2763975236UL); /* 37 */
    HH ( d, a, b, c, in[ 4], S32, 1272893353UL); /* 38 */
    HH ( c, d, a, b, in[ 7], S33, 4139469664UL); /* 39 */
    HH ( b, c, d, a, in[10], S34, 3200236656UL); /* 40 */
    HH ( a, b, c, d, in[13], S31,  681279174UL); /* 41 */
    HH ( d, a, b, c, in[ 0], S32, 3936430074UL); /* 42 */
    HH ( c, d, a, b, in[ 3], S33, 3572445317UL); /* 43 */
    HH ( b, c, d, a, in[ 6], S34,   76029189UL); /* 44 */
    HH ( a, b, c, d, in[ 9], S31, 3654602809UL); /* 45 */
    HH ( d, a, b, c, in[12], S32, 3873151461UL); /* 46 */
    HH ( c, d, a, b, in[15], S33,  530742520UL); /* 47 */
    HH ( b, c, d, a, in[ 2], S34, 3299628645UL); /* 48 */

    /* Round 4 */
#define S41 6
#define S42 10
#define S43 15
#define S44 21
    II ( a, b, c, d, in[ 0], S41, 4096336452UL); /* 49 */
    II ( d, a, b, c, in[ 7], S42, 1126891415UL); /* 50 */
    II ( c, d, a, b, in[14], S43, 2878612391UL); /* 51 */
    II ( b, c, d, a, in[ 5], S44, 4237533241UL); /* 52 */
    II ( a, b, c, d, in[12], S41, 1700485571UL); /* 53 */
    II ( d, a, b, c, in[ 3], S42, 2399980690UL); /* 54 */
    II ( c, d, a, b, in[10], S43, 4293915773UL); /* 55 */
    II ( b, c, d, a, in[ 1], S44, 2240044497UL); /* 56 */
    II ( a, b, c, d, in[ 8], S41, 1873313359UL); /* 57 */
    II ( d, a, b, c, in[15], S42, 4264355552UL); /* 58 */
    II ( c, d, a, b, in[ 6], S43, 2734768916UL); /* 59 */
    II ( b, c, d, a, in[13], S44, 1309151649UL); /* 60 */
    II ( a, b, c, d, in[ 4], S41, 4149444226UL); /* 61 */
    II ( d, a, b, c, in[11], S42, 3174756917UL); /* 62 */
    II ( c, d, a, b, in[ 2], S43,  718787259UL); /* 63 */
    II ( b, c, d, a, in[ 9], S44, 3951481745UL); /* 64 */

    buf[0] += a;
    buf[1] += b;
    buf[2] += c;
    buf[3] += d;
}

/* ------------------------------------------------------------- */


/*********************************************/
/** Original md5.c modified from now on ~CP **/
/*********************************************/

/* Prints a md5digest in human readable representation */
void MDPrintArr(FILE * stream, unsigned char *digest)
{
    for(int i = 0; i < 16; i++)
    {
        fprintf(stream,"%02x", digest[i]);
    }
}
